### DISCUSSION TITLE
Datastore outage and consumer retry and restart query

### DISCUSSION THREAD

**Discussion by cookie-bytes** [BEagle1984/silverback#210] [Q&A] [answered]
In the event of a data store outage, our aim is to have our service consumers reattempt processing events. Instead of allowing messages to be transferred to a Dead Letter Queue (DLQ), we prefer to maintain the consumer offset at the current position.

We are considering setting up a retry policy that will attempt reprocessing ‘x’ number of times. Once these attempts have been exhausted, Silverback will halt the consumer.

To monitor the status of the consumer, we will implement health checks using the Consumer Management API.

If the consumer is not running, we will restart the service and re-enter the retry loop. Once the data store is back online, the service will resume processing events successfully.

Does the above sound sensible? Is there a more suitable approach available?

Thanks

**Comment by cookie-bytes** [comment]
K8s will be responsible for the health check and restarting the services.

**Comment by BEagle1984** [comment]
Yes, that would work.

You could also just let Silverback retry indefinitely (e.g. with increasingly longer intervals), to let it recover without manual interaction.

**Reply by cookie-bytes** [reply]
What is the best way to configure your retry indefinitely approach?

Thank you for all your help

**Reply by BEagle1984** [reply, marked as answer]
It will simply retry indefinitely if you configure a retry policy without specifying the `retriesCount`: see [IErrorPolicyBuilder](https://silverback-messaging.net/api/Silverback.Messaging.Configuration.IErrorPolicyBuilder.html#Silverback_Messaging_Configuration_IErrorPolicyBuilder_Retry_System_Action_Silverback_Messaging_Inbound_ErrorHandling_RetryErrorPolicy__).

Note that you can build sophisticated policies like this:
```csharp
.OnError(policy => policy
    .Retry(1) // 1st retry is immediate
    .ThenRetry(3, TimeSpan.FromSeconds(10)) // 2nd-4th attempts 10 seconds apart
    .ThenRetry(TimeSpan.FromMinutes(2))) // then retry indefinitely every 2 minutes
```

Remember to increase the [MaxPollIntervalMs](https://silverback-messaging.net/api/Silverback.Messaging.Configuration.Kafka.ConfluentConsumerConfigProxy.html#Silverback_Messaging_Configuration_Kafka_ConfluentConsumerConfigProxy_MaxPollIntervalMs), if you want to increase the delay past the default 5 minutes.

**Reply by cookie-bytes** [reply]
Thank you as always for great insight
